src/state_machine.zig

```diff
@@ -324,29 +310,49 @@ pub const StateMachine = struct {
         assert(t.timestamp > dr.timestamp);
         assert(t.timestamp > cr.timestamp);
 
-        assert(!t.flags.auto_commit);
-        if (dr.debit_reserved < t.amount) return .debit_amount_was_not_reserved;
-        if (cr.credit_reserved < t.amount) return .credit_amount_was_not_reserved;
+        assert(t.flags.two_phase_commit);
+        if (dr.debits_reserved < t.amount) return .debit_amount_was_not_reserved;
+        if (cr.credits_reserved < t.amount) return .credit_amount_was_not_reserved;
 
-        if (dr.exceeds_debit_accepted_limit(t.amount)) return .exceeds_debit_accepted_limit;
-        if (dr.exceeds_credit_accepted_limit(t.amount)) return .exceeds_credit_accepted_limit;
+        // TODO Should we check limits again here?
+        // On the one hand, it's possible for a subsequent transfer to change balances.
+        // On the other hand, the spirit of two-phase commit is that we reserve resources upfront.
```

Comment:
@adrianhopebailie we only check `debits_must_not_exceed_credits` when reserving resources for a two-phase transfer, i.e. during `create_transfers` but not in `commit_transfers`.

Three strategies:

1. We keep to the spirit of two-phase commit, doing all resource reservation during the prepare, and allow concurrent transfers of credits out of the account (after the prepare, but before the fulfill) to invalidate the limit invariant (this is the current behavior now).
2. We check the limits again here when committing, however concurrent transfers of credits out of the account (after the prepare, but before the fulfill) could then fail the transfer even when it was prepared successfully.
3. If `debits_must_not_exceed_credits` is set, then we fail any concurrent transfers of credits out of the account (after the prepare, but before the fulfill) that would prevent prepared transfers from committing.

Alternatively

Reply:
I've worked through this again, and clarified my thinking and resolved this with an assert in 0325d945263e2cf656ff9ebee695243f890fc66d:

We don't need the second check in `commit_transfers` because it should be impossible for concurrent transfers of credits out of the account to break the same limit we're checking, so that we were in fact already doing option 3, which has to be the right option.